### PR TITLE
Update Jedi calls for its 0.17.0+ API

### LINKED ISSUE #744: Adapt to the last Jedi
Hi,
Error happend with fresh installed python-language-server.
After check the error message, it cause by incompatable dependency that python-language-server depend on 0.14<=JEDI<0.16, while the pip actually install the last JEDI (0.16.0) when install pyls.

The workaround is change the setup.py to depend on JEDI 0.15.2.

Is there any plan to cachup with last JEDI (0.16.*)?

Thanks

The JEDI releases are:
https://github.com/davidhalter/jedi/releases

### PR DESCRIPTION
**Pull request by bnavigator**
As discussed in https://github.com/palantir/python-language-server/issues/744#issuecomment-614338490, fixes #744, including most recent Jedi 0.17.0

~This is a combination of #775 and dalthviz#1~ Edit: removed the jedi master branch pulls in the CIs

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
         'configparser; python_version<"3.0"',
         'future>=0.14.0; python_version<"3"',
         'backports.functools_lru_cache; python_version<"3.2"',
-        'jedi>=0.14.1,<0.16',
+        'jedi>=0.16.0',
```
Comment (ccordoba12) [inline setup.py]: Please change this to

    'jedi>=0.17.0,<0.18.0'

because:

* Jedi 0.16 was giving errors for Pandas.
* Jedi is changing a lot lately, so before it reaches 1.0, we don't want to have a `>=` on it here.

**Review by ccordoba12 — COMMENTED** [review]
@bnavigator, as part of this PR please also remove the code that tried to support Jedi 0.14 because it no longer applies (please look for JEDI_VERSION to do do that).

File: setup.py (hunk 1 above)
Comment (bnavigator) [reply setup.py]: Done. ~The CI failures don't look related.~

Comment (bnavigator) [reply setup.py]: Ooops.

A comma was missing between jedi and json-rpcserver line.

**Comment by ccordoba12** [comment]
Thanks a lot @bnavigator for your help with this one! I'll review it tomorrow (I don't have time today, sorry).

**Comment by goanpeca** [comment]
Thanks a lot for this @bnavigator!!

**Comment by goanpeca** [comment]
IS this ready @ccordoba12 ?

**Comment by E1k3** [comment]
Is there any update on this or is there any way to help finalizing this pull request?

**Comment by ccordoba12** [comment]
> IS this ready @ccordoba12 ?

I haven't had time to review it. Please help me with that and let me know if you need additional help.

File: pyls/plugins/jedi_completion.py [hunk 2]
```diff
@@ -173,7 +176,7 @@ def _label(definition):
 def _detail(definition):
     try:
         return definition.parent().full_name or ''
-    except AttributeError:
+    except (AttributeError, TypeError):
```
Comment (E1k3) [inline pyls/plugins/jedi_completion.py]: Why is TypeError necessary now and was not before? `_detail` is always called with `definition` being a `Completion`, which always implements `parent()`.

**Review by E1k3 — APPROVED** [review]
The changes seem good to me. The changes from https://github.com/davidhalter/jedi/issues/1063#issuecomment-578457199 are accounted for and the new `Project` feature of jedi 0.17 is used.

I will now test this branch with vim + asyncomplete + vim-lsp for definition, hover, completion, references and signature. As far as I know, highlights are not implemented for this client, yet.

**Comment by E1k3** [comment]
I have tested all mentioned features in a relatively simple script that imports various modules from Keras, NumPy, SciPy, Argparse and others and pyls worked flawlessly.

@ccordoba12  If more comprehensive testing or more detailed commentary on the changes are needed, please tell me in detail what you need..

File: pyls/plugins/jedi_completion.py (hunk 2 above)
Comment (bnavigator) [reply pyls/plugins/jedi_completion.py]: You're right. This is a remnant from cb51b09254efac0b2c22130815198fab487db54f when `definition` was not a completion.

**Review by E1k3 — APPROVED** [review]

**Comment by E1k3** [comment]
@ccordoba12 Is there anything to be done with this request despite merging it?

The archlinux package of pyls is currently incompatible with jedi and workarounds are on hold, because this looks finished.

**Comment by bnavigator** [comment]
@E1k3, I recommend you apply the patch directly to the Archlinux package. That's [what OpenSUSE is doing](https://build.opensuse.org/package/show/openSUSE:Factory/python-python-language-server).

There may be some burnt ground between @ccordoba12 and me. Chances are, he will never accept this PR despite the clear demand from the community shown by the reactions in this thread and in #744.

**Comment by goanpeca** [comment]
@bnavigator we really appreciate your help with this work. We are a bit crowded right now and in the midst of making a release. Having this PR merged could delay our release or introduce some unwanted bugs with some other libraries introspection.

>There may be some burnt ground between @ccordoba12 and me. Chances are, he will never accept this PR despite the clear demand from the community shown by the reactions in this thread and in #744.

We just need to review the PR (I have not found time yet!)

We also probably need to add some simple smoke tests with libraries that have broken with the past versions of Jedi (0.16).

Thanks again for the help!

**Comment by ccordoba12** [comment]
> Chances are, he will never accept this PR despite the clear demand from the community shown by the reactions in this thread and in #744.

This is not true. As @goanpeca said, we are swamped at the moment and haven't had time to review this PR, sorry. We'll try to do it and release 0.32.0 with these changes this week.

**Comment by ccordoba12** [comment]
> We also probably need to add some simple smoke tests with libraries that have broken with the past versions of Jedi (0.16).

I already added tests for Numpy, Pandas and Matplotlib. One of them detected an error in Jedi 0.16, which was fixed in 0.17

File: pyls/_utils.py [hunk 3]
```diff
@@ -152,6 +149,17 @@ def clip_column(column, lines, line_number):
     return min(column, max_column)
 
 
+def position_to_jedi_linecolumn(document, position):
+    """Convert the format 'line', 'character' to 'line', 'column'"""
+    code_position = {}
+    if position:
```
Comment (goanpeca) [inline pyls/_utils.py:162]: @bnavigator what are the possible values for position?

File: pyls/plugins/hover.py [hunk 4]
```diff
@@ -9,43 +9,37 @@
...
 
+    # raw docstring returns only doc, without signature
+    doc = _utils.format_docstring(definition.docstring(raw=True))
+
+    # Find first exact matching signature
+    signature = next((x.to_string() for x in definition.get_signatures()
+                      if x.name == word), '')
+
+    contents = []
+    if signature:
```
Comment (goanpeca) [inline pyls/plugins/hover.py:36]: To make it more readable could we add spaces between if statements?

like:

```python
    if signature:
        contents.append({
            'language': 'python',
            'value': signature,
        })

    if doc:
        contents.append(doc)

    if not contents:
        return {'contents': ''}
```

File: pyls/plugins/jedi_completion.py [hunk 5]
```diff
@@ -51,16 +51,18 @@
 @hookimpl
 def pyls_completions(config, document, position):
     try:
-        definitions = document.jedi_script(position).completions()
+        code_position = _utils.position_to_jedi_linecolumn(document, position)
```
Comment (goanpeca) [inline pyls/plugins/jedi_completion.py]: Is this something that could fail?

If not we could move it out of the try?

File: pyls/plugins/jedi_completion.py [hunk 6]
```diff
@@ -138,9 +140,9 @@ def _format_completion(d, include_params=True):
         path = path.replace('/', '\\/')
         completion['insertText'] = path
 
-    if (include_params and hasattr(d, 'params') and d.params and
-            not is_exception_class(d.name)):
-        positional_args = [param for param in d.params
+    sig = d.get_signatures()
```
Comment (goanpeca) [inline pyls/plugins/jedi_completion.py:135]: Why was the

`hasattr(d, 'params')`no longer needed?

File: pyls/plugins/jedi_completion.py [hunk 7]
```diff
@@ -138,9 +140,9 @@ def _format_completion(d, include_params=True):
...
+    if (include_params and sig and not is_exception_class(d.name)):
+        positional_args = [param for param in sig[0].params
```
Comment (goanpeca) [inline pyls/plugins/jedi_completion.py:137]: is `sig`guaranteed to have at least the [0] item?

File: pyls/workspace.py [hunk 8]
```diff
@@ -213,16 +213,9 @@ def word_at_position(self, position):
         return m_start[0] + m_end[-1]
 
     def jedi_names(self, all_scopes=False, definitions=True, references=False):
-        environment_path = None
```
Comment (goanpeca) [inline pyls/workspace.py:216]: Why was all this code removed?

**Review by goanpeca — CHANGES_REQUESTED** [review]
Hi @bnavigator sorry this took long. Minor suggestions and some questions to better understand the changes.

Thanks!

File: pyls/plugins/jedi_completion.py (hunk 5 above)
Comment (E1k3) [reply pyls/plugins/jedi_completion.py]: Should be moved out, since the catch is only there for missing CompiledObject attributes.

As stated in line /60, the `try..catch ` block may not be necessary at all. But until this is investigated, the `try..catch ` should stay

File: pyls/plugins/jedi_completion.py (hunk 6 above)
Comment (E1k3) [reply pyls/plugins/jedi_completion.py:135]: `get_signatures` always returns a list of `BaseSignature`s, which always have the attribute `params` (see https://jedi.readthedocs.io/en/latest/docs/api-classes.html?highlight=get_signatures#jedi.api.classes.BaseName.get_signatures)

File: pyls/workspace.py (hunk 8 above)
Comment (E1k3) [reply pyls/workspace.py:216]: This was redundant code (see line 229/222). `jedi.Script(...).get_names()` uses the environment specified in `jedi.Script(environment=...)`.

File: pyls/plugins/jedi_completion.py (hunk 7 above)
Comment (E1k3) [reply pyls/plugins/jedi_completion.py:137]: Empty `sig` lists don't fullfill the if condition `if ... and sig:`. If you want it more explicit, change it to `... and len(sig) > 0`.

**Review by E1k3 — APPROVED** [review]
I hope I've answered some of your questions

Unfortunately the jedi documentation pre 0.16 is gone, so finding out how the deprecated/removed functions worked is not straight forward. This should be easier for those that wrote the code that was changed in this PR.

Based on the current jedi documentation, these changes look sound.

**Comment by bnavigator** [comment]
Thanky you @E1k3 for answering these questions!

File: pyls/_utils.py (hunk 3 above)
Comment (bnavigator) [reply pyls/_utils.py:162]: PyLS's wideline used position dict for the LSP protocol
```python
position = {'line': int, 'character': int}
```

https://microsoft.github.io/language-server-protocol/specification#position

File: pyls/plugins/jedi_completion.py (hunk 5 above)
Comment (bnavigator) [reply pyls/plugins/jedi_completion.py]: The `try...catch` can be removed. Fixed upstream https://github.com/davidhalter/jedi/issues/1452

**Review by E1k3 — APPROVED** [review]

File: pyls/plugins/jedi_completion.py (hunk 6 above)
Comment (goanpeca) [reply pyls/plugins/jedi_completion.py:135]: Awesome, thanks!

File: pyls/plugins/jedi_completion.py (hunk 7 above)
Comment (goanpeca) [reply pyls/plugins/jedi_completion.py:137]: Right!, I missed that :-)

File: pyls/workspace.py (hunk 8 above)
Comment (goanpeca) [reply pyls/workspace.py:216]: Thanks!

**Review by goanpeca — APPROVED** [review]
Thanks @bnavigator for the work and @E1k3 for the comments.

File: pyls/workspace.py [hunk 9]
```diff
@@ -237,15 +230,16 @@ def jedi_script(self, position=None):
         environment = self.get_enviroment(environment_path) if environment_path else None
 
         kwargs = {
-            'source': self.source,
+            'code': self.source,
             'path': self.path,
-            'sys_path': sys_path,
             'environment': environment,
+            'project': jedi.api.Project(os.path.dirname(self.path),
```
Comment (fsouza) [inline pyls/workspace.py]: Should this use `self._workspace.root_path`? See #801.

File: pyls/_utils.py [hunk 10]
```diff
@@ -140,18 +139,32 @@ def format_docstring(contents):
...
     contents = contents.replace('  ', u'\u00A0' * 2)
-    if LooseVersion(JEDI_VERSION) < LooseVersion('0.15.0'):
-        contents = contents.replace('*', '\\*')
     return contents
 
 
 def clip_column(column, lines, line_number):
-    # Normalise the position as per the LSP that accepts character positions > line length
-    # https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#position
+    """Normalise the position as per the LSP that accepts character positions > line length
```
Comment (ccordoba12) [inline pyls/_utils.py]:
```suggestion
    """
    Normalise the position as per the LSP that accepts character positions > line length
```

File: pyls/_utils.py [hunk 11]
```diff
@@ -140,18 +139,32 @@ def format_docstring(contents):
...
+
+    https://microsoft.github.io/language-server-protocol/specification#position
+    """
     max_column = len(lines[line_number].rstrip('\r\n')) if len(lines) > line_number else 0
     return min(column, max_column)
 
 
+def position_to_jedi_linecolumn(document, position):
+    """Convert the LSP format 'line', 'character' to Jedi's 'line', 'column'
```
Comment (ccordoba12) [inline pyls/_utils.py]:
```suggestion
    """
    Convert the LSP format 'line', 'character' to Jedi's 'line', 'column'
```

File: pyls/plugins/jedi_completion.py [hunk 12]
```diff
@@ -50,17 +50,11 @@
...
 def pyls_completions(config, document, position):
-    try:
-        definitions = document.jedi_script(position).completions()
-    except AttributeError as e:
-        if 'CompiledObject' in str(e):
-            # Needed to handle missing CompiledObject attribute
-            # 'sub_modules_dict'
-            definitions = None
-        else:
-            raise e
```
Comment (ccordoba12) [inline pyls/plugins/jedi_completion.py:61]: Why did you remove this try/except?

File: pyls/workspace.py [hunk 13]
```diff
@@ -237,15 +230,16 @@ def jedi_script(self, position=None):
...
+                                        sys_path=sys_path),
         }
 
         if position:
-            kwargs['line'] = position['line'] + 1
-            kwargs['column'] = _utils.clip_column(position['character'], self.lines, position['line'])
+            # deprecated by Jedi to use in Script() constructor
```
Comment (ccordoba12) [inline pyls/workspace.py]:
```suggestion
            # Deprecated by Jedi to use in Script() constructor
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot @bnavigator for your help with this! I left some minor style suggestions and a question for you.

Please also answer the question @fsouza made yesterday.

File: pyls/plugins/jedi_completion.py (hunk 12 above)
Comment (bnavigator) [reply pyls/plugins/jedi_completion.py:61]: Already answered. https://github.com/palantir/python-language-server/pull/781#discussion_r422069504

File: pyls/workspace.py (hunk 9 above)
Comment (bnavigator) [reply pyls/workspace.py]: @ccordoba12 wants me to answer this question. But I think a PyLS developer should know this better than me. So should it? I don't know. What difference does it make? How is the internal _workspace attribute used in PyLS?

File: pyls/_utils.py (hunk 10 above)
Comment (bnavigator) [reply pyls/_utils.py]: Very picky considering how the rest of the file looks. PEP 257 allows both. And don't suggest stuff, if you can't wait and push the commit by yourself a few minutes later. We have been waiting for your review for weeks.

**Comment by bnavigator** [comment]
https://github.com/krassowski/jupyterlab-lsp/issues/270#issuecomment-626134634

Another testimony of how badly this is needed and how pinning stuff to old versions causes problems.

File: pyls/workspace.py (hunk 9 above)
Comment (ccordoba12) [reply pyls/workspace.py]: After taking a look at this, I think @fsouza is right.

Please make this change @bnavigator.

**Comment by ccordoba12** [comment]
> Another testimony of how badly this is needed and how pinning stuff to old versions causes problems.

That's Jedi's fault by not declaring its dependency on Parso correctly (and please don't try to start another discussion about this).

File: pyls/workspace.py [hunk 14]
```diff
@@ -237,15 +230,16 @@ def jedi_script(self, position=None):
         environment = self.get_enviroment(environment_path) if environment_path else None
```
Comment (bnavigator) [inline pyls/workspace.py:232]:
```suggestion

        project_path = self._workspace.root_path if self._workspace else os.path.dirname(self.path)
```

File: pyls/workspace.py (hunk 9 above)
Comment (bnavigator) [reply pyls/workspace.py]:
```suggestion
            'project': jedi.Project(path=project_path, sys_path=sys_path),
```

**Review by bnavigator — COMMENTED** [review]
Taken from #801

File: pyls/workspace.py [hunk 15]
```diff
@@ -236,16 +229,18 @@ def jedi_script(self, position=None):
...
 
+        project_path = self._workspace.root_path if self._workspace else os.path.dirname(self.path)
         kwargs = {
-            'source': self.source,
+            'code': self.source,
             'path': self.path,
-            'sys_path': sys_path,
             'environment': environment,
+            'project': jedi.Project(path=project_path, sys_path=sys_path),
+                                        sys_path=sys_path),
```
Comment (ccordoba12) [inline pyls/workspace.py]: There's a clear syntax error here.

File: pyls/workspace.py [hunk 16]
```diff
@@ -229,11 +229,12 @@ def jedi_script(self, position=None):
...
         environment = self.get_enviroment(environment_path) if environment_path else None
 
+        project_path = self._workspace.root_path if self._workspace else os.path.dirname(self.path)
         kwargs = {
             'code': self.source,
             'path': self.path,
             'environment': environment,
-            'project': jedi.api.Project(os.path.dirname(self.path),
+            'project': jedi.Project(path=project_path, sys_path=sys_path),
                                         sys_path=sys_path),
```
Comment (bnavigator) [inline pyls/workspace.py]:
```suggestion
            'project': jedi.Project(path=project_path, sys_path=sys_path),
```

File: pyls/workspace.py (hunk 15 above)
Comment (bnavigator) [reply pyls/workspace.py]: On it. Leave it alone for a few minutes.

Comment (ccordoba12) [reply pyls/workspace.py]: Ok

**Comment by ccordoba12** [comment]
> That's Jedi's fault by not declaring its dependency on Parso correctly (and please don't try to start another discussion about this).

Sorry, it wasn't. It's an error in Jedi that's fixed already (still I don't want to talk more about it, we're doing our job here trying to get this merged).

**Comment by bnavigator** [comment]
Using _workspace.root_path didn't look like such a great idea.

```
>       project_path = self._workspace.root_path if self._workspace else os.path.dirname(self.path)
E       AttributeError: 'MockWorkspace' object has no attribute 'root_path'
```

**Comment by ccordoba12** [comment]
Ok, on it. I think I know how to fix this.

**Comment by ccordoba12** [comment]
I don't think your last commit is the right solution @bnavigator. I think the problem is in `MockWorkspace`.

Could you leave this alone for a few minutes? Thanks!

**Comment by bnavigator** [comment]
If you allow `Document(..., workspace=None)` you cannot just assume that `._workspace` is a workspace.

**Comment by ccordoba12** [comment]
> If you allow Document(..., workspace=None) you cannot just assume that ._workspace is a workspace.

I think it's better to use something that stands for a real `Workspace` object (I'm almost ready with my fixes).

**Comment by fsouza** [comment]
@ccordoba12 I think what @bnavigator means is that now that workspace is required in Document, it shouldn't be an optional parameter in the constructor. As long as it's optional in the constructor, the code should support the case where it's `None`, even though in practice this never happens? 🤔

**Review by ccordoba12 — DISMISSED** [review]
I think this is finally ready. Thanks everyone (specially @bnavigator) for your patience.

**Comment by ccordoba12** [comment]
You're totally @bnavigator and @fsouza! I'll do that last change here.

**Comment by bnavigator** [comment]
> @ccordoba12 I think what @bnavigator means is that now that workspace is required in Document, it shouldn't be an optional parameter in the constructor. As long as it's optional in the constructor, the code should support the case where it's `None`, even though in practice this never happens?

Thanks for breaking it down @fsouza.

At least in the tests it happens all the time. And these are only the one-liners:

```
[ben@voyagerS9:~/src/python-language-server]% grep 'Document(.*)' * -R | grep -v workspace                                     [0]
test/fixtures.py:    return Document(DOC_URI, DOC)
test/plugins/test_autopep8_format.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_autopep8_format.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_autopep8_format.py:    doc = Document(DOC_URI, GOOD_DOC)
test/plugins/test_flake8_lint.py:    doc = Document(uris.from_fs_path(name))
test/plugins/test_flake8_lint.py:    doc = Document('', DOC)
test/plugins/test_folding.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_folding.py:    doc = Document(DOC_URI, SYNTAX_ERR)
test/plugins/test_mccabe_lint.py:        doc = Document(DOC_URI, DOC)
test/plugins/test_mccabe_lint.py:    doc = Document(DOC_URI, DOC_SYNTAX_ERR)
test/plugins/test_pycodestyle_lint.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_pydocstyle_lint.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_pydocstyle_lint.py:    doc = Document(TEST_DOC_URI, "")
test/plugins/test_pydocstyle_lint.py:    doc = Document(DOC_URI, "")
test/plugins/test_pydocstyle_lint.py:    doc = Document(DOC_URI, "bad syntax")
test/plugins/test_pyflakes_lint.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_pyflakes_lint.py:    doc = Document(DOC_URI, DOC_SYNTAX_ERR)
test/plugins/test_pyflakes_lint.py:    doc = Document(DOC_URI, DOC_UNDEFINED_NAME_ERR)
test/plugins/test_pyflakes_lint.py:    doc = Document(DOC_URI, DOC_ENCODING)
test/plugins/test_pylint_lint.py:        yield Document(uris.from_fs_path(name))
test/plugins/test_pylint_lint.py:        config, Document(uris.from_fs_path(__file__)), True)
test/plugins/test_pylint_lint.py:        config, Document(uris.from_fs_path(__file__)), False)
test/plugins/test_yapf_format.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_yapf_format.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_yapf_format.py:    doc = Document(DOC_URI, GOOD_DOC)
test/plugins/test_yapf_format.py:    doc = Document(uris.from_fs_path(src.strpath), DOC)
test/plugins/test_completion.py:    doc = Document(DOC_URI, DOC)
test/plugins/test_rope_rename.py:    doc = Document(DOC_URI)
test/test_document.py:    document_mem = Document(DOC_URI, u'my source')
test/test_document.py:    document_disk = Document(DOC_URI)
test/test_document.py:    doc = Document('file:///uri', u'')
test/test_document.py:    doc = Document('file:///uri', u'itshelloworld')
test/test_document.py:    doc = Document('file:///uri', u''.join(old))
test/test_document.py:    doc = Document('file:///uri', u''.join(old))
```

**Review by fsouza — APPROVED** [review]
🎉

**Review by ccordoba12 — APPROVED** [review]

**Comment by mattst88** [comment]
Thanks a bunch to everyone involved!

If possible, please tag a release with this included. It'll help downstream Linux distros quite a lot.

**Comment by polyzen** [comment]
https://github.com/palantir/python-language-server/milestone/17

**Comment by Foxboron** [comment]
Thanks for the work getting this fixed :)

File: pyls/workspace.py [hunk 17]
```diff
@@ -112,8 +112,8 @@ def _create_document(self, doc_uri, source=None, version=None):
 
 class Document(object):
 
-    def __init__(self, uri, source=None, version=None, local=True, extra_sys_path=None, rope_project_builder=None,
-                 config=None, workspace=None):
+    def __init__(self, uri, workspace, source=None, version=None, local=True, extra_sys_path=None,
+                 rope_project_builder=None, config=None):
```
Comment (bnavigator) [inline pyls/workspace.py:116]: Is it guaranteed that `Document()`  is not instantiated directly by downstream projects or did you just change the API, so to speak?

Comment (ccordoba12) [reply pyls/workspace.py:116]: I think it's not possible to have a document without a workspace and that that was the case before only to simplify the tests.

However, I took a quick look at the most important third-party plugins of this project (pyls-my and pyls-black) and they don't seem to use `Document` directly.

**Comment by ccordoba12** [comment]
0.32.0 is out.
